File: Difficulty/Medium/[127] Word Ladder/solution.cpp
/*
 * @lc app=leetcode id=127 lang=cpp
 *
 * [127] Word Ladder
 */

// @lc code=start
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        if (find(wordList.begin(), wordList.end(), endWord) == wordList.end()) return 0;
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        vector<string> visited(1, beginWord);
        int level = 1, res = 0, visit_begin = 0;
        queue<string> q;
        q.push(beginWord);
        while (!q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++){
                string curr = q.front(); q.pop();
                for (int j = visit_begin; j < visited.size(); ++j)
                    wordSet.erase(visited[j]);
                visit_begin = visited.size();
                addNeighboursToQ(curr, wordSet, visited, q, endWord, level, res);
                if (res != 0) return res;
            }
            level++;
        }
        return 0;
    }

    void addNeighboursToQ(string& curr, unordered_set<string>& wordSet, vector<string>& visited, queue<string>& q, const string& endWord, int& level, int& res){
        for (string s: wordSet){
            if (isOneCharDiff(curr, s)){
                q.push(s);
                visited.push_back(s);
                if (s == endWord) {
                    res = level + 1;
                    return;
                }
            }
        }
    }

    bool isOneCharDiff(const string& s1, const string& s2){
        int num = 0;
        for (int i = 0; i < s1.size(); i++){
            if (s1[i] != s2[i]) num++;
            if (num > 1) return false;
        }
        return num == 1;
    }
};
// @lc code=end

File: Difficulty/Medium/[127] Word Ladder/solution.py
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList: return 0
        wordDict, visited = dict(), []
        for w in wordList:
            wordDict[w] = 0
        if wordDict.get(beginWord) is not None: wordDict.pop(beginWord)
        level, res, visited_begin = 1, 0, 0
        q = deque()
        q.append(beginWord)
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                for j in range(visited_begin, len(visited)):
                    wordDict.pop(visited[j])
                visited_begin = len(visited)    
                res = self.addNeighboursToQ(curr, wordDict, visited, q, endWord, level)
                if res != 0: return res
            level += 1
        return 0
    
    def addNeighboursToQ(self, curr, wordSet, visited, q, endWord, level):
        for s in wordSet:
            if self.isOneCharDiff(curr, s):
                q.append(s)
                visited.append(s)
                if s == endWord:
                    return level + 1
        return 0

    def isOneCharDiff(self, s1, s2):
        num = 0
        for i in range(len(s1)):
            if s1[i] != s2[i]: num += 1
            if num > 1: return False
        return num == 1